// @primer/react/lib-esm/experimental/IssueLabel/IssueLabel.js
import { c } from 'react-compiler-runtime';
import { clsx } from 'clsx';
import { useTheme } from '../../ThemeProvider.js';
import { getColorsFromHex } from './getColorFromHex.js';
import classes from './IssueLabel.module.css.js';
import { jsx } from 'react/jsx-runtime';

function IssueLabel(t0) {
  const $ = c(42);
  let children;
  let className;
  let fillColor;
  let props;
  let style;
  let t1;
  if ($[0] !== t0) {
    ({
      children,
      className,
      fillColor,
      style,
      variant: t1,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = className;
    $[3] = fillColor;
    $[4] = props;
    $[5] = style;
    $[6] = t1;
  } else {
    children = $[1];
    className = $[2];
    fillColor = $[3];
    props = $[4];
    style = $[5];
    t1 = $[6];
  }
  const variant = t1 === undefined ? "gray" : t1;
  const {
    resolvedColorScheme
  } = useTheme();
  const mode = resolvedColorScheme !== null && resolvedColorScheme !== void 0 && resolvedColorScheme.startsWith("dark") ? "dark" : "light";
  let t2;
  let t3;
  let t4;
  if ($[7] !== className || $[8] !== fillColor || $[9] !== mode || $[10] !== resolvedColorScheme || $[11] !== style || $[12] !== variant) {
    const bgColors = {
      light: "#ffffff",
      dark: "#0d1117"
    };
    if ($[16] !== className) {
      t2 = clsx(className, classes.IssueLabel);
      $[16] = className;
      $[17] = t2;
    } else {
      t2 = $[17];
    }
    t3 = fillColor ? undefined : variant;
    t4 = fillColor ? {
      ...style,
      ...getColorsFromHex(fillColor, resolvedColorScheme, bgColors[mode])
    } : style;
    $[7] = className;
    $[8] = fillColor;
    $[9] = mode;
    $[10] = resolvedColorScheme;
    $[11] = style;
    $[12] = variant;
    $[13] = t2;
    $[14] = t3;
    $[15] = t4;
  } else {
    t2 = $[13];
    t3 = $[14];
    t4 = $[15];
  }
  let t5;
  if ($[18] !== t2 || $[19] !== t3 || $[20] !== t4) {
    t5 = {
      className: t2,
      "data-variant": t3,
      style: t4
    };
    $[18] = t2;
    $[19] = t3;
    $[20] = t4;
    $[21] = t5;
  } else {
    t5 = $[21];
  }
  const sharedProps = t5;
  if ("as" in props && props.as) {
    let BaseComponent;
    let rest;
    if ($[22] !== props) {
      ({
        as: BaseComponent,
        ...rest
      } = props);
      $[22] = props;
      $[23] = BaseComponent;
      $[24] = rest;
    } else {
      BaseComponent = $[23];
      rest = $[24];
    }
    let t6;
    if ($[25] !== BaseComponent || $[26] !== children || $[27] !== rest || $[28] !== sharedProps) {
      t6 = /*#__PURE__*/jsx(BaseComponent, {
        ...rest,
        ...sharedProps,
        children: children
      });
      $[25] = BaseComponent;
      $[26] = children;
      $[27] = rest;
      $[28] = sharedProps;
      $[29] = t6;
    } else {
      t6 = $[29];
    }
    return t6;
  }
  if ("href" in props) {
    let t6;
    if ($[30] !== children || $[31] !== props || $[32] !== sharedProps) {
      t6 = /*#__PURE__*/jsx("a", {
        ...props,
        ...sharedProps,
        children: children
      });
      $[30] = children;
      $[31] = props;
      $[32] = sharedProps;
      $[33] = t6;
    } else {
      t6 = $[33];
    }
    return t6;
  }
  if ("onClick" in props) {
    let t6;
    if ($[34] !== children || $[35] !== props || $[36] !== sharedProps) {
      t6 = /*#__PURE__*/jsx("button", {
        type: "button",
        ...props,
        ...sharedProps,
        children: children
      });
      $[34] = children;
      $[35] = props;
      $[36] = sharedProps;
      $[37] = t6;
    } else {
      t6 = $[37];
    }
    return t6;
  }
  let t6;
  if ($[38] !== children || $[39] !== props || $[40] !== sharedProps) {
    t6 = /*#__PURE__*/jsx("span", {
      ...props,
      ...sharedProps,
      children: children
    });
    $[38] = children;
    $[39] = props;
    $[40] = sharedProps;
    $[41] = t6;
  } else {
    t6 = $[41];
  }
  return t6;
}

export { IssueLabel };
